feat(general): add reporting contributor metrics

maxamel opened the pull request:
This PR adds capability to report contributor metrics in case we are in prisma context. They are reported by running git log in case of CI integration and sending them to external API.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

gruebel commented on checkov/main.py:
```diff
@@ -415,6 +425,40 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
+        process = subprocess.Popen(['git', 'shortlog', '-ne', '--all', '--since', '"90 days ago"', '--pretty="commit-%ct"', '--reverse'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)  # nosec
+        out, err = process.communicate()
+        if err:
+            logger.info('Failed to collect contributor metrics')
```
can you also log the `err` itself here

gruebel commented on checkov/main.py:
```diff
@@ -415,6 +425,40 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
 
+def report_contributor_metrics(repository: str) -> None:  # ignore: type
+
+    def _parse_gitlog() -> dict[str, Any] | None:
+        process = subprocess.Popen(['git', 'shortlog', '-ne', '--all', '--since', '"90 days ago"', '--pretty="commit-%ct"', '--reverse'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)  # nosec
+        out, err = process.communicate()
+        if err:
+            logger.info('Failed to collect contributor metrics')
+            return None
+        # split per contributor
+        list_of_contributors = out.decode('utf-8').split('\n\n')
```
it would nice to move this into a normal function, then it easy to mock it and then write some UTs with expected output and what should come out at the end.

maxamel replied:
Done

Eliran-Turgeman commented on checkov/main.py:
```diff
@@ -415,6 +425,40 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
+        splittedList = contributor.split('\n')
+        user = splittedList[0]
+        commit = splittedList[1]
+        return user[0:user.find('(')] + commit[commit.find('-') + 1:len(commit)].strip('\"')
```
The `0` and `len(commit)` can be omitted

```suggestion
        return user[:user.find('(')] + commit[commit.find('-') + 1:].strip('\"')
```

Eliran-Turgeman commented on checkov/main.py:
```diff
@@ -415,6 +425,40 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
+            "POST", f"{bc_integration.api_url}/api/v1/contributors/report",
+            headers=bc_integration.get_default_headers("POST"), data=json.dumps(request_body)
+        )
+        if response.status_code < 300:
```
Just a suggestion, you can use `response.ok` instead (it checks if the status code is below 400)

maxamel replied:
I actually used it in the beginning but then I noticed it marks anything below 400 as ok.
I would expect only 200/202/204 to be an OK response.

nimrodkor commented on checkov/main.py:
```diff
@@ -415,6 +425,40 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
+def report_contributor_metrics(repository: str) -> None:  # ignore: type
+
+    def _parse_gitlog() -> dict[str, Any] | None:
+        process = subprocess.Popen(['git', 'shortlog', '-ne', '--all', '--since', '"90 days ago"', '--pretty="commit-%ct"', '--reverse'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)  # nosec
```
@maxamel Can we use gitpython instead of subprocess?
https://pypi.org/project/GitPython/

gruebel replied:
I also wanted to suggest it, but it looks like that `shortlog` doesn't work very well with `gitpython` https://github.com/gitpython-developers/GitPython/discussions/1434

nimrodkor commented on checkov/main.py:
```diff
@@ -249,6 +251,14 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
                                                         repo_branch=config.branch,
                                                         prisma_api_url=config.prisma_api_url)
+
+            if source.name in [BCSourceType.GITHUB_ACTIONS, BCSourceType.JENKINS, BCSourceType.CIRCLECI, BCSourceType.CODEBUILD] \
```
@maxamel Maybe we should add to the `BCSource` enum another field, `is_ci`, and have it by default `False` but for these 4 have it as `True`?

maxamel replied:
Done

nimrodkor reviewed (commented):
Very nice @maxamel

Eliran-Turgeman commented on checkov/main.py:
```diff
@@ -415,6 +425,40 @@ def run(banner: str = checkov_banner, argv: List[str] = sys.argv[1:]) -> Optiona
     return None
 
 
+def report_contributor_metrics(repository: str) -> None:  # ignore: type
```
I think it might be worth extracting it to a different module, IMO it's not too related to the main script

maxamel commented:
Extracted reporting logic to separate file, added UTs, extended BcSource to include collaborator reporting flag, and added env variable protection on this flow.

nimrodkor reviewed (approved):
I think this is more concise, but it is a style issue. Approved anyway @maxamel

maxamel commented on checkov/common/bridgecrew/bc_source.py:
```diff
@@ -24,16 +25,16 @@ class BCSourceType:
+    BCSourceType.CODEBUILD: SourceType(BCSourceType.CODEBUILD, True, True),
+    BCSourceType.JENKINS: SourceType(BCSourceType.JENKINS, True, True),
+    BCSourceType.CIRCLECI: SourceType(BCSourceType.CIRCLECI, True, True),
+    BCSourceType.ADMISSION_CONTROLLER: SourceType(BCSourceType.ADMISSION_CONTROLLER, False, False)
```
Cool fixed
